File: backend/app/models/aitracker.py
"""Models for the v2 AI Answer-Engine Visibility Tracker (PRODUCT_SPEC §8).

Scheduled, managed, tier-gated tracking on top of the one-shot GEO scans:
org-scoped tracker configs (named prompt sets + target market/language +
competitors + refresh cadence), per-run parsed answers, Visibility Score /
Share-of-Voice rollups with run-over-run deltas, and the Mention-Gap report.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import Field

from app.models.common import AppModel
from app.models.geo import Citation, SentimentLabel


class TrackerEngine(str, Enum):
    """Answer engines the v2 tracker queries.

    This is the tracker's own enum (per spec: ChatGPT, Perplexity, Gemini,
    Google AI Overviews, Google AI Mode). It extends beyond
    :class:`app.models.geo.AnswerEngine`, which lacks ``google_ai_mode``.
    """

    CHATGPT = "chatgpt"
    PERPLEXITY = "perplexity"
    GEMINI = "gemini"
    GOOGLE_AI_OVERVIEWS = "google_ai_overviews"
    GOOGLE_AI_MODE = "google_ai_mode"


class RefreshCadence(str, Enum):
    """How often a tracker config is refreshed (plan-gated)."""

    DAILY = "daily"
    WEEKLY = "weekly"


# ---------------------------------------------------------------------------
# Tracker configs (managed prompt sets)
# ---------------------------------------------------------------------------


class TrackerConfigCreate(AppModel):
    name: str
    brand: str
    prompts: list[str] = []
    competitors: list[str] = []
    engines: list[TrackerEngine] = []  # empty -> the plan's full allowed engine set
    market: str = "global"
    language: str = "en"
    own_domain: str | None = None  # None -> derived from the brand name
    refresh_cadence: RefreshCadence = RefreshCadence.WEEKLY


class TrackerConfigUpdate(AppModel):
    """Partial update; ``None`` means "keep the current value"."""

    name: str | None = None
    brand: str | None = None
    prompts: list[str] | None = None
    competitors: list[str] | None = None
    engines: list[TrackerEngine] | None = None  # [] -> reset to the plan default
    market: str | None = None
    language: str | None = None
    own_domain: str | None = None
    refresh_cadence: RefreshCadence | None = None


class TrackerConfig(AppModel):
    id: str
    name: str
    brand: str
    prompts: list[str]
    competitors: list[str] = []
    engines: list[TrackerEngine]
    market: str = "global"
    language: str = "en"
    own_domain: str
    refresh_cadence: RefreshCadence = RefreshCadence.WEEKLY
    created_at: datetime
    last_run_at: datetime | None = None


class DueConfig(AppModel):
    """A config due for a scheduled refresh (returned by the ops endpoint)."""

    config_id: str
    name: str
    brand: str
    refresh_cadence: RefreshCadence
    last_run_at: datetime | None = None  # None -> never run


# ---------------------------------------------------------------------------
# Parsed answers
# ---------------------------------------------------------------------------


class TrackerAnswer(AppModel):
    """One engine answer for one prompt, parsed for brand/competitor signals."""

    engine: TrackerEngine
    prompt: str
    text: str = ""
    ranked_list: list[str] = []
    citations: list[Citation] = []
    brand_mentioned: bool = False
    brand_mentions: int = 0  # whole-phrase text mentions
    brand_position: int | None = None  # 1-based position within the answer's ranked list
    competitors_mentioned: list[str] = []  # sorted, distinct
    sentiment_label: SentimentLabel = SentimentLabel.NEUTRAL
    sentiment_score: float = Field(default=0.0, ge=-1.0, le=1.0)


# ---------------------------------------------------------------------------
# Metrics rollups
# ---------------------------------------------------------------------------


class RollupMetrics(AppModel):
    """Aggregates for one engine (or overall). See ``app.core.aitracker.engine``
    for the exact deterministic formulas."""

    prompts_total: int = 0
    prompts_mentioned: int = 0
    mention_rate: float = 0.0  # prompts_mentioned / prompts_total
    avg_position: float | None = None  # mean 1-based rank where present (1 = best)
    position_score: float = Field(default=0.0, ge=0.0, le=1.0)  # mean of 1/rank (0 if absent)
    citation_share: float = Field(default=0.0, ge=0.0, le=1.0)  # own-domain / all citations
    visibility_score: float = Field(default=0.0, ge=0.0, le=100.0)  # weighted blend, 0-100
    share_of_voice: float = 0.0  # brand mentions / (brand + competitor mentions)
    competitor_share_of_voice: dict[str, float] = {}
    avg_sentiment: float = Field(default=0.0, ge=-1.0, le=1.0)
    sentiment_label: SentimentLabel = SentimentLabel.NEUTRAL
    # Run-over-run trend deltas (None on the first run).
    visibility_score_delta: float | None = None
    share_of_voice_delta: float | None = None
    mention_rate_delta: float | None = None
    avg_position_delta: float | None = None  # positive = improvement (closer to #1)


class EngineRollup(RollupMetrics):
    engine: TrackerEngine


class VisibilityRollup(AppModel):
    """The per-run metrics rollup (latest one served by GET .../visibility)."""

    config_id: str
    brand: str
    run_at: datetime
    engines: list[EngineRollup] = []
    overall: RollupMetrics


# ---------------------------------------------------------------------------
# Mention-Gap report
# ---------------------------------------------------------------------------


class MentionGapEntry(AppModel):
    """A prompt where competitors are mentioned but the brand is not."""

    prompt: str
    gap_engines: list[TrackerEngine] = []  # engines where the gap occurs
    competitors_mentioned: list[str] = []  # sorted, distinct across gap engines
    opportunity_score: float = 0.0  # gap engines x distinct competitors


class MentionGapReport(AppModel):
    config_id: str
    brand: str
    run_at: datetime
    total_prompts: int = 0
    entries: list[MentionGapEntry] = []  # ranked by opportunity_score desc


# ---------------------------------------------------------------------------
# Run report
# ---------------------------------------------------------------------------


class TrackerRunReport(AppModel):
    """Everything one tracker run produced: answers, rollup, mention gap."""

    config_id: str
    brand: str
    run_at: datetime
    engines: list[TrackerEngine] = []
    answers: list[TrackerAnswer] = []
    rollup: VisibilityRollup
    mention_gap: MentionGapReport

File: mcp-server/src/tools/aitracker.ts
/**
 * AI Answer-Engine Visibility Tracker MCP tools (v2, PRODUCT_SPEC §8).
 *
 * Scheduled, managed, tier-gated tracking on top of the one-shot GEO scans:
 * org-scoped tracker configs (named prompt sets + market/language +
 * competitors + refresh cadence), on-demand runs, Visibility Score /
 * Share-of-Voice rollups with run-over-run deltas, and the Mention-Gap
 * report — backed by the backend's /api/v1/ai-tracker/* routes. Every route
 * requires the `ai_tracker` plan feature.
 */

import { z } from "zod";
import { apiClient } from "../apiClient.js";
import { buildQuery, defineTool, type ToolDefinition } from "./define.js";

/** Answer engines the v2 tracker queries (models/aitracker.py TrackerEngine). */
const TRACKER_ENGINES = [
  "chatgpt",
  "perplexity",
  "gemini",
  "google_ai_overviews",
  "google_ai_mode",
] as const;

/** Plan-gated refresh cadences (models/aitracker.py RefreshCadence). */
const REFRESH_CADENCES = ["daily", "weekly"] as const;

const trackerEngineEnum = z.enum(TRACKER_ENGINES);
const refreshCadenceEnum = z.enum(REFRESH_CADENCES);

const trackerEnginesJsonSchema = {
  type: "array",
  description: "Answer engines to track (empty/omitted = the plan's full allowed engine set)",
  items: { type: "string", enum: [...TRACKER_ENGINES] },
} as const;

const configIdJsonSchema = {
  type: "string",
  description: "Tracker config ID",
} as const;

const configIdSchema = z.string().min(1, "config_id is required");

function configPath(configId: string, suffix = ""): string {
  return `/ai-tracker/configs/${encodeURIComponent(configId)}${suffix}`;
}

// ---------------------------------------------------------------------------
// tracker_create_config — POST /ai-tracker/configs
// ---------------------------------------------------------------------------

const trackerCreateConfigSchema = z.object({
  name: z.string().min(1, "name is required"),
  brand: z.string().min(1, "brand is required"),
  prompts: z.array(z.string().min(1)).optional(),
  competitors: z.array(z.string().min(1)).optional(),
  engines: z.array(trackerEngineEnum).optional(),
  market: z.string().optional(),
  language: z.string().optional(),
  own_domain: z.string().optional(),
  refresh_cadence: refreshCadenceEnum.optional(),
});

const trackerCreateConfigTool = defineTool(
  "tracker_create_config",
  "Create an AI visibility tracker config: a named prompt set for a brand with target " +
    "market/language, competitors, answer engines, and a daily/weekly refresh cadence " +
    "(plan-gated). Returns the stored config including its ID.",
  trackerCreateConfigSchema,
  {
    type: "object",
    properties: {
      name: { type: "string", description: "Human-readable name for the tracker config" },
      brand: { type: "string", description: "Brand name to track visibility for" },
      prompts: {
        type: "array",
        description: "Prompts to track the brand against",
        items: { type: "string" },
      },
      competitors: {
        type: "array",
        description: "Competitor brand names for share-of-voice comparison",
        items: { type: "string" },
      },
      engines: trackerEnginesJsonSchema,
      market: { type: "string", description: "Target market, e.g. global, US, DE (default global)" },
      language: { type: "string", description: "Prompt language code (default en)" },
      own_domain: { type: "string", description: "The brand's own domain (omitted = derived from the brand name)" },
      refresh_cadence: { type: "string", enum: [...REFRESH_CADENCES], description: "Scheduled refresh cadence (default weekly; plan-gated)" },
    },
    required: ["name", "brand"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/ai-tracker/configs", args)
);

// ---------------------------------------------------------------------------
// tracker_list_configs — GET /ai-tracker/configs
// ---------------------------------------------------------------------------

const trackerListConfigsSchema = z.object({});

const trackerListConfigsTool = defineTool(
  "tracker_list_configs",
  "List all of the organization's AI visibility tracker configs with their prompt sets, " +
    "engines, cadence, and last run time.",
  trackerListConfigsSchema,
  {
    type: "object",
    properties: {},
    required: [],
    additionalProperties: false,
  },
  () => apiClient.get("/ai-tracker/configs")
);

// ---------------------------------------------------------------------------
// tracker_get_config — GET /ai-tracker/configs/{config_id}
// ---------------------------------------------------------------------------

const trackerGetConfigSchema = z.object({
  config_id: configIdSchema,
});

const trackerGetConfigTool = defineTool(
  "tracker_get_config",
  "Get a single AI visibility tracker config by ID: prompts, competitors, engines, " +
    "market/language, own domain, refresh cadence, and last run time.",
  trackerGetConfigSchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.get(configPath(args.config_id))
);

// ---------------------------------------------------------------------------
// tracker_update_config — PUT /ai-tracker/configs/{config_id}
// ---------------------------------------------------------------------------

const trackerUpdateConfigSchema = z.object({
  config_id: configIdSchema,
  name: z.string().min(1).optional(),
  brand: z.string().min(1).optional(),
  prompts: z.array(z.string().min(1)).optional(),
  competitors: z.array(z.string().min(1)).optional(),
  engines: z.array(trackerEngineEnum).optional(),
  market: z.string().optional(),
  language: z.string().optional(),
  own_domain: z.string().optional(),
  refresh_cadence: refreshCadenceEnum.optional(),
});

const trackerUpdateConfigTool = defineTool(
  "tracker_update_config",
  "Partially update an AI visibility tracker config: only the supplied fields change " +
    "(an empty engines array resets to the plan default). Returns the updated config.",
  trackerUpdateConfigSchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
      name: { type: "string", description: "New name for the tracker config" },
      brand: { type: "string", description: "New brand name to track" },
      prompts: {
        type: "array",
        description: "Replacement prompt set",
        items: { type: "string" },
      },
      competitors: {
        type: "array",
        description: "Replacement competitor list",
        items: { type: "string" },
      },
      engines: trackerEnginesJsonSchema,
      market: { type: "string", description: "New target market" },
      language: { type: "string", description: "New prompt language code" },
      own_domain: { type: "string", description: "New own domain" },
      refresh_cadence: { type: "string", enum: [...REFRESH_CADENCES], description: "New refresh cadence (plan-gated)" },
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  ({ config_id, ...body }) => apiClient.put(configPath(config_id), body)
);

// ---------------------------------------------------------------------------
// tracker_delete_config — DELETE /ai-tracker/configs/{config_id}
// ---------------------------------------------------------------------------

const trackerDeleteConfigSchema = z.object({
  config_id: configIdSchema,
});

const trackerDeleteConfigTool = defineTool(
  "tracker_delete_config",
  "Delete an AI visibility tracker config (and its stored runs) by ID.",
  trackerDeleteConfigSchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.delete(configPath(args.config_id))
);

// ---------------------------------------------------------------------------
// tracker_run — POST /ai-tracker/configs/{config_id}/run
// ---------------------------------------------------------------------------

const trackerRunSchema = z.object({
  config_id: configIdSchema,
  now: z.string().optional(),
});

const trackerRunTool = defineTool(
  "tracker_run",
  "Run an AI visibility tracker config now: query every configured engine with every " +
    "prompt and return the full run report — parsed answers, the Visibility Score / " +
    "Share-of-Voice rollup (with run-over-run deltas), and the Mention-Gap report.",
  trackerRunSchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
      now: { type: "string", description: "Optional ISO-8601 timestamp to record as the run time (defaults to now)" },
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.post(`${configPath(args.config_id, "/run")}${buildQuery({ now: args.now })}`, {})
);

// ---------------------------------------------------------------------------
// tracker_history — GET /ai-tracker/configs/{config_id}/history
// ---------------------------------------------------------------------------

const trackerHistorySchema = z.object({
  config_id: configIdSchema,
});

const trackerHistoryTool = defineTool(
  "tracker_history",
  "Get the run-over-run history of a tracker config: one Visibility Score / Share-of-Voice " +
    "rollup per run, in chronological order, for trend charting.",
  trackerHistorySchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.get(configPath(args.config_id, "/history"))
);

// ---------------------------------------------------------------------------
// tracker_visibility — GET /ai-tracker/configs/{config_id}/visibility
// ---------------------------------------------------------------------------

const trackerVisibilitySchema = z.object({
  config_id: configIdSchema,
});

const trackerVisibilityTool = defineTool(
  "tracker_visibility",
  "Get the latest visibility rollup for a tracker config: per-engine and overall mention " +
    "rate, average position, citation share, 0-100 Visibility Score, Share of Voice vs " +
    "competitors, sentiment, and deltas vs the previous run. 404 if the config has no runs yet.",
  trackerVisibilitySchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.get(configPath(args.config_id, "/visibility"))
);

// ---------------------------------------------------------------------------
// tracker_mention_gap — GET /ai-tracker/configs/{config_id}/mention-gap
// ---------------------------------------------------------------------------

const trackerMentionGapSchema = z.object({
  config_id: configIdSchema,
});

const trackerMentionGapTool = defineTool(
  "tracker_mention_gap",
  "Get the latest Mention-Gap report for a tracker config: prompts where competitors are " +
    "mentioned but the brand is not, with the gap engines and an opportunity score per " +
    "prompt, ranked by opportunity. 404 if the config has no runs yet.",
  trackerMentionGapSchema,
  {
    type: "object",
    properties: {
      config_id: configIdJsonSchema,
    },
    required: ["config_id"],
    additionalProperties: false,
  },
  (args) => apiClient.get(configPath(args.config_id, "/mention-gap"))
);

// ---------------------------------------------------------------------------
// tracker_due — GET /ai-tracker/due
// ---------------------------------------------------------------------------

const trackerDueSchema = z.object({
  now: z.string().optional(),
});

const trackerDueTool = defineTool(
  "tracker_due",
  "Ops endpoint (admin/owner only): list the org's tracker configs due for a scheduled " +
    "refresh based on their cadence and last run time — then trigger each with tracker_run.",
  trackerDueSchema,
  {
    type: "object",
    properties: {
      now: { type: "string", description: "Optional ISO-8601 timestamp to evaluate due-ness against (defaults to now)" },
    },
    required: [],
    additionalProperties: false,
  },
  (args) => apiClient.get(`/ai-tracker/due${buildQuery({ now: args.now })}`)
);

// ---------------------------------------------------------------------------

/** All AI Answer-Engine Visibility Tracker tools exposed by this MCP server. */
export const aitrackerTools: ToolDefinition[] = [
  trackerCreateConfigTool,
  trackerListConfigsTool,
  trackerGetConfigTool,
  trackerUpdateConfigTool,
  trackerDeleteConfigTool,
  trackerRunTool,
  trackerHistoryTool,
  trackerVisibilityTool,
  trackerMentionGapTool,
  trackerDueTool,
];
